--- a/include/configs/mvebu_armada-37xx.h
+++ b/include/configs/mvebu_armada-37xx.h
@@ -111,11 +111,22 @@
 #include <config_distro_bootcmd.h>
 
 #define CONFIG_EXTRA_ENV_SETTINGS	\
+	"bootcmd=run emmc_boot\0"	\
 	"scriptaddr=0x4d00000\0"	\
 	"pxefile_addr_r=0x4e00000\0"	\
 	"fdt_addr_r=0x4f00000\0"	\
 	"kernel_addr_r=0x5000000\0"	\
 	"ramdisk_addr_r=0x8000000\0"	\
+	"image_name=Image\0"		\
+	"emmc_dev=/dev/mmcblk0\0"	\
+	"fdt_name=uDPU.dtb\0"		\
+	"initramfs_name=openwrt-mvebu-cortexa53-uDPU-initramfs.itb\0" \
+	"console=console=ttyMV0,115200 earlycon=ar3700_uart,0xd0012000\0"	\
+	"bootargs_emmc=setenv bootargs $console root=${emmc_dev}p3 rw rootwait\0"	\
+	"emmc_boot=mmc dev 0; ext4load mmc 0:1 $kernel_addr_r $image_name; "	\
+		"ext4load mmc 0:1 $fdt_addr_r $fdt_name; run bootargs_emmc; "	\
+		"booti $kernel_addr_r - $fdt_addr_r\0"	\
+	"stage3=setenv bootargs $console; tftpboot $kernel_addr_r $initramfs_name; bootm $kernel_addr_r\0" \
 	BOOTENV
 
 #endif /* _CONFIG_MVEBU_ARMADA_37XX_H */
