--- a/drivers/net/phy/sfp.c
+++ b/drivers/net/phy/sfp.c
@@ -153,7 +153,7 @@ static const enum gpiod_flags gpio_flags
  */
 #define T_PROBE_INIT	msecs_to_jiffies(300)
 #define T_HPOWER_LEVEL	msecs_to_jiffies(300)
-#define T_PROBE_RETRY	msecs_to_jiffies(100)
+#define T_PROBE_RETRY	msecs_to_jiffies(6000)
 
 /* SFP modules appear to always have their PHY configured for bus address
  * 0x56 (which with mdio-i2c, translates to a PHY address of 22).
